### ISSUE TITLE
Fatal Bus Error in FMRIPREP

### ISSUE THREAD

**Issue by sulantha2006** [poldracklab/fmriprep#1555]
I am runnign fmriprep through singularity. Image was created on 19th March 2019, using tag latest. 

When I am running a single subject: I get a fatal Bus Error. 

Here is the -vvv debug log when the error occurs. 

```
No resources available
190320-18:01:14,880 nipype.workflow DEBUG:
	 No resources available
190320-18:01:16,878 nipype.workflow DEBUG:
	 No resources available
190320-18:01:18,880 nipype.workflow DEBUG:
	 No resources available
190320-18:01:20,883 nipype.workflow DEBUG:
	 No resources available
190320-18:01:22,407 nipype.workflow INFO:
	 [Node] Finished "_inu_n40".
190320-18:01:22,412 nipype.workflow INFO:
	 [Node] Finished "fmriprep_wf.single_subject_NI9226IM20181130_wf.anat_preproc_wf.brain_extraction_wf.inu_n4".
190320-18:01:22,883 nipype.workflow INFO:
	 [Job 69] Completed (fmriprep_wf.single_subject_NI9226IM20181130_wf.anat_preproc_wf.brain_extraction_wf.inu_n4).
190320-18:01:22,884 nipype.workflow DEBUG:
	 Tasks currently running: 0. Pending: 0.
190320-18:01:22,886 nipype.workflow INFO:
	 [MultiProc] Running 0 tasks, and 6 jobs ready. Free memory (GB): 226.62/226.62, Free processors: 1/1.
190320-18:01:22,978 nipype.workflow DEBUG:
	 Allocating fmriprep_wf.single_subject_NI9226IM20181130_wf.anat_preproc_wf.anat_derivatives_wf.ds_t1_preproc ID=70 (0.20GB, 1 threads). Free: 226.42GB, 0 threads.
190320-18:01:22,978 nipype.workflow DEBUG:
	 Setting node inputs
190320-18:01:22,979 nipype.workflow DEBUG:
	 input: source_file
190320-18:01:22,979 nipype.workflow DEBUG:
	 results file: /grave/fmriprep_wf/single_subject_NI9226IM20181130_wf/anat_preproc_wf/anat_derivatives_wf/t1_name/result_t1_name.pklz
190320-18:01:22,979 nipype.utils DEBUG:
	 Loading pkl: /grave/fmriprep_wf/single_subject_NI9226IM20181130_wf/anat_preproc_wf/anat_derivatives_wf/t1_name/result_t1_name.pklz
190320-18:01:22,980 nipype.workflow DEBUG:
	 output: out
190320-18:01:22,980 nipype.workflow DEBUG:
	 [Node] ds_t1_preproc - setting input source_file = /data/data05/Studies/MCSA/RAW/MRIBIDS/sub-NI9226IM20181130/ses-1/anat/sub-NI9226IM20181130_T1w.nii.gz
190320-18:01:22,980 nipype.workflow DEBUG:
	 input: in_file
190320-18:01:22,980 nipype.workflow DEBUG:
	 results file: /grave/fmriprep_wf/single_subject_NI9226IM20181130_wf/anat_preproc_wf/brain_extraction_wf/inu_n4/result_inu_n4.pklz
190320-18:01:22,980 nipype.utils DEBUG:
	 Loading pkl: /grave/fmriprep_wf/single_subject_NI9226IM20181130_wf/anat_preproc_wf/brain_extraction_wf/inu_n4/result_inu_n4.pklz
190320-18:01:22,981 nipype.workflow DEBUG:
	 output: output_image
190320-18:01:22,981 nipype.workflow DEBUG:
	 [Node] ds_t1_preproc - setting input in_file = ['/grave/fmriprep_wf/single_subject_NI9226IM20181130_wf/anat_preproc_wf/brain_extraction_wf/inu_n4/mapflow/_inu_n40/sub-NI9226IM20181130_ses-1_T1w_maths_corrected.nii']
190320-18:01:22,982 nipype.workflow DEBUG:
	 [Node] Directory not found "/grave/fmriprep_wf/single_subject_NI9226IM20181130_wf/anat_preproc_wf/anat_derivatives_wf/ds_t1_preproc".
190320-18:01:22,982 nipype.workflow DEBUG:
	 Checking hash "fmriprep_wf.single_subject_NI9226IM20181130_wf.anat_preproc_wf.anat_derivatives_wf.ds_t1_preproc" locally: cached=False, updated=False.
190320-18:01:22,982 nipype.workflow DEBUG:
	 Running node fmriprep_wf.single_subject_NI9226IM20181130_wf.anat_preproc_wf.anat_derivatives_wf.ds_t1_preproc on master thread
190320-18:01:22,982 nipype.workflow INFO:
	 [Node] Setting-up "fmriprep_wf.single_subject_NI9226IM20181130_wf.anat_preproc_wf.anat_derivatives_wf.ds_t1_preproc" in "/grave/fmriprep_wf/single_subject_NI9226IM20181130_wf/anat_preproc_wf/anat_derivatives_wf/ds_t1_preproc".
190320-18:01:22,982 nipype.workflow DEBUG:
	 [Node] Directory not found "/grave/fmriprep_wf/single_subject_NI9226IM20181130_wf/anat_preproc_wf/anat_derivatives_wf/ds_t1_preproc".
190320-18:01:22,982 nipype.utils DEBUG:
	 Removing contents of /grave/fmriprep_wf/single_subject_NI9226IM20181130_wf/anat_preproc_wf/anat_derivatives_wf/ds_t1_preproc
190320-18:01:22,982 nipype.utils DEBUG:
	 Creating directory /grave/fmriprep_wf/single_subject_NI9226IM20181130_wf/anat_preproc_wf/anat_derivatives_wf/ds_t1_preproc
190320-18:01:23,653 nipype.utils DEBUG:
	 Creating directory /grave/fmriprep_wf/single_subject_NI9226IM20181130_wf/anat_preproc_wf/anat_derivatives_wf/ds_t1_preproc/_report
190320-18:01:24,45 nipype.workflow DEBUG:
	 [Node] Writing pre-exec report to "/grave/fmriprep_wf/single_subject_NI9226IM20181130_wf/anat_preproc_wf/anat_derivatives_wf/ds_t1_preproc/_report/report.rst"
190320-18:01:25,638 nipype.workflow INFO:
	 [Node] Running "ds_t1_preproc" ("smriprep.interfaces.DerivativesDataSink")
190320-18:01:26,386 nipype.utils DEBUG:
	 /data/data05/Studies/MCSA/PROCESSED/FMRI/FMRIPREP/fmriprep/sub-NI9226IM20181130/anat/sub-NI9226IM20181130_desc-preproc_T1w.nii
190320-18:01:26,540 nipype.utils DEBUG:
	 Linking File: /data/data05/Studies/MCSA/PROCESSED/FMRI/FMRIPREP/fmriprep/sub-NI9226IM20181130/anat/sub-NI9226IM20181130_desc-preproc_T1w.nii->/grave/fmriprep_wf/single_subject_NI9226IM20181130_wf/anat_preproc_wf/brain_extraction_wf/inu_n4/mapflow/_inu_n40/sub-NI9226IM20181130_ses-1_T1w_maths_corrected.nii
190320-18:01:26,829 nipype.utils DEBUG:
	 Copying File: /data/data05/Studies/MCSA/PROCESSED/FMRI/FMRIPREP/fmriprep/sub-NI9226IM20181130/anat/sub-NI9226IM20181130_desc-preproc_T1w.nii->/grave/fmriprep_wf/single_subject_NI9226IM20181130_wf/anat_preproc_wf/brain_extraction_wf/inu_n4/mapflow/_inu_n40/sub-NI9226IM20181130_ses-1_T1w_maths_corrected.nii
Fatal Python error: Bus error

Thread 0x00007fc04e2d9700 (most recent call first):
  File "/usr/local/miniconda/lib/python3.7/threading.py", line 296 in wait
  File "/usr/local/miniconda/lib/python3.7/multiprocessing/queues.py", line 224 in _feed
  File "/usr/local/miniconda/lib/python3.7/threading.py", line 865 in run
  File "/usr/local/miniconda/lib/python3.7/threading.py", line 917 in _bootstrap_inner
  File "/usr/local/miniconda/lib/python3.7/threading.py", line 885 in _bootstrap

Thread 0x00007fc04eada700 (most recent call first):
  File "/usr/local/miniconda/lib/python3.7/selectors.py", line 415 in select
  File "/usr/local/miniconda/lib/python3.7/multiprocessing/connection.py", line 920 in wait
  File "/usr/local/miniconda/lib/python3.7/concurrent/futures/process.py", line 354 in _queue_management_worker
  File "/usr/local/miniconda/lib/python3.7/threading.py", line 865 in run
  File "/usr/local/miniconda/lib/python3.7/threading.py", line 917 in _bootstrap_inner
  File "/usr/local/miniconda/lib/python3.7/threading.py", line 885 in _bootstrap

Current thread 0x00007fc0766e7700 (most recent call first):
  File "/usr/local/miniconda/lib/python3.7/site-packages/nibabel/volumeutils.py", line 833 in _write_data
  File "/usr/local/miniconda/lib/python3.7/site-packages/nibabel/volumeutils.py", line 660 in array_to_file
  File "/usr/local/miniconda/lib/python3.7/site-packages/nibabel/arraywriters.py", line 562 in to_fileobj
  File "/usr/local/miniconda/lib/python3.7/site-packages/nibabel/analyze.py", line 1090 in to_file_map
  File "/usr/local/miniconda/lib/python3.7/site-packages/nibabel/filebasedimages.py", line 334 in to_filename
  File "/usr/local/miniconda/lib/python3.7/site-packages/niworkflows/interfaces/bids.py", line 431 in _run_interface
  File "/usr/local/miniconda/lib/python3.7/site-packages/nipype/interfaces/base/core.py", line 375 in run
  File "/usr/local/miniconda/lib/python3.7/site-packages/nipype/pipeline/engine/nodes.py", line 637 in _run_command
  File "/usr/local/miniconda/lib/python3.7/site-packages/nipype/pipeline/engine/nodes.py", line 557 in _run_interface
  File "/usr/local/miniconda/lib/python3.7/site-packages/nipype/pipeline/engine/nodes.py", line 473 in run
  File "/usr/local/miniconda/lib/python3.7/site-packages/nipype/pipeline/plugins/multiproc.py", line 316 in _send_procs_to_workers
  File "/usr/local/miniconda/lib/python3.7/site-packages/nipype/pipeline/plugins/base.py", line 183 in run
  File "/usr/local/miniconda/lib/python3.7/site-packages/nipype/pipeline/engine/workflows.py", line 599 in run
  File "/usr/local/miniconda/lib/python3.7/site-packages/fmriprep/cli/run.py", line 436 in main
  File "/usr/local/miniconda/bin/fmriprep", line 11 in <module>
/var/lib/gridengine/default/spool/alzheimer/job_scripts/109044: line 11:  4051 Bus error 


```

**Comment by oesteban** [comment]
Hi @sulantha2006, this looks like a system failure. Have you contacted support for your cluster?

**Comment by sulantha2006** [comment]
Hi @oesteban , I am not sure. I actually compressed all my nii files to nii.gz and it got fixed, I know this is a cheap workaround, but I guess this is still related to the mmap issue of nipype.

**Comment by oesteban** [comment]
I see, you are right: it seems like a problem related to mmap. Thanks for posting the solution, we'll dig deeper into it if some other users run into this (and gzipping does not work for them).

**Comment by boegel** [comment]
I'm running into a similar `Bus error`, looks like it could be the same issue.

@sulantha2006 @oesteban: Do you have any more information about this nipype mmap issue? Is there a workaround? Which nii files did you gzip exactly to dance around this? (I'm clueless about the use of FMRIprep and the data files it requires, I'm just trying to benchmark it...)

**Comment by sulantha2006** [comment]
@boegel I gzed my T1 files and the Bold files. I would say, gz all the nii files.

**Comment by boegel** [comment]
As far as I can tell, all the `*.nii` files in the input data I'm using is already gzipped, they're all in ` sub-*/func/` or `sub-*/anat/` subdirectories, does that sound right?

```
$ find ds009 -name '*.nii*' | grep -v nii.gz
$ find ds009 -name '*.nii*' | grep nii.gz | wc -l
192
$ find ds009 -name '*.nii*' | grep /func/ | wc -l
144
$ find ds009 -name '*.nii*' | grep /anat/ | wc -l
48
```

**Comment by sulantha2006** [comment]
@boegel I am not an expert to advice on this. But can you post the fmriprep command you use and  debug log (option -vvv). 
That may give you a better idea.

**Comment by thomascope** [comment]
> I see, you are right: it seems like a problem related to mmap. Thanks for posting the solution, we'll dig deeper into it if some other users run into this (and gzipping does not work for them).

Just posting because I'm having the same issue. My images aren't gzipped - I will try that.

**Comment by ds7711** [comment]
@oesteban It looks like I am running into the same issues. I ran the same dataset both using singularity on the cluster and docker on a local desktop. The errors from both are almost identical. Using fMRIprep 1.3.2. Will try the "gz" solution.

**Comment by jeroenvanbaar** [comment]
Just wanted to add here that I ran into the same issue, gzipped my Niftis, problem solved! Easy step to implement when renaming/reorganizing files for BIDS format.

**Comment by effigies** [comment]
I'm going to reopen this because I've noticed a couple things:

First, this can be fixed in [niworkflows.interfaces.bids.DerivativesDataSink._run_interface](https://github.com/poldracklab/niworkflows/blob/254f4b4fcc5e6ecb29d2f4602a30786b913ecce5/niworkflows/interfaces/bids.py#L470) by adding `mmap=False` to the `nb.load` call.

Second, we seem to be saving uncompressed NIfTI-1 images in derivatives:

```
Copying File: /data/data05/Studies/MCSA/PROCESSED/FMRI/FMRIPREP/fmriprep/sub-NI9226IM20181130/anat/sub-NI9226IM20181130_desc-preproc_T1w.nii->/grave/fmriprep_wf/single_subject_NI9226IM20181130_wf/anat_preproc_wf/brain_extraction_wf/inu_n4/mapflow/_inu_n40/sub-NI9226IM20181130_ses-1_T1w_maths_corrected.nii
```

@oesteban Is that desired or a bug? My guess is that we tried to avoid compressing CIFTI-2 and accidentally disabled compressing NIfTI-1.

**Comment by effigies** [comment]
Also, just in general principle, we should support all valid BIDS datasets with BOLD data, and making people compress their inputs isn't very user friendly.

**Comment by marloula** [comment]
We receive the same error but I am not sure if it is the same as the one mentioned by Sulantha. We work in linux. see output in the attachment. I did select a working directory. 
![Foto 05-06-19 om 14 05](https://user-images.githubusercontent.com/51357763/59034276-0186cb00-886b-11e9-9fd6-9168488115df.jpg)

**Comment by effigies** [comment]
Hi @marloula I don't see a bus error in your outputs, but it looks like you have a pre-existing FreeSurfer directory for `sub-01` with no data in it. If you want fMRIPrep to populate it, just remove that directory.

**Comment by marloula** [comment]
Dear Effigies,

Thank you for your response. however there is a bus error and I can not get fmriprep to work properly it does seem to convert al functional images to Bids_spilt images.nii.gz however I do not get a quality overview report, based upon which I am assuming it does not work fully yet.

this is the latest error I received:
[Node] Finished "fmriprep_wf.single_subject_03_wf.func_preproc_task_taste3_run_3_echo_1_wf.sdc_wf.sdc_unwarp_wf.unwarp_reference".
Fatal Python error: Bus error

Thread 0x00007f904d4fe700 (most recent call first):
  File "/usr/local/miniconda/lib/python3.7/threading.py", line 296 in wait
  File "/usr/local/miniconda/lib/python3.7/multiprocessing/queues.py", line 224 in _feed
  File "/usr/local/miniconda/lib/python3.7/threading.py", line 865 in run
  File "/usr/local/miniconda/lib/python3.7/site-packages/sentry_sdk/integrations/threading.py", line 45 in run
  File "/usr/local/miniconda/lib/python3.7/threading.py", line 917 in _bootstrap_inner
  File "/usr/local/miniconda/lib/python3.7/threading.py", line 885 in _bootstrap

Thread 0x00007f904dcff700 (most recent call first):
  File "/usr/local/miniconda/lib/python3.7/selectors.py", line 415 in select
  File "/usr/local/miniconda/lib/python3.7/multiprocessing/connection.py", line 920 in wait
  File "/usr/local/miniconda/lib/python3.7/concurrent/futures/process.py", line 354 in _queue_management_worker
  File "/usr/local/miniconda/lib/python3.7/threading.py", line 865 in run
  File "/usr/local/miniconda/lib/python3.7/site-packages/sentry_sdk/integrations/threading.py", line 45 in run
  File "/usr/local/miniconda/lib/python3.7/threading.py", line 917 in _bootstrap_inner
  File "/usr/local/miniconda/lib/python3.7/threading.py", line 885 in _bootstrap

Thread 0x00007f9056f16700 (most recent call first):
  File "/usr/local/miniconda/lib/python3.7/threading.py", line 296 in wait
  File "/usr/local/miniconda/lib/python3.7/queue.py", line 170 in get
  File "/usr/local/miniconda/lib/python3.7/site-packages/sentry_sdk/worker.py", line 102 in _target
  File "/usr/local/miniconda/lib/python3.7/threading.py", line 865 in run
  File "/usr/local/miniconda/lib/python3.7/site-packages/sentry_sdk/integrations/threading.py", line 45 in run
  File "/usr/local/miniconda/lib/python3.7/threading.py", line 917 in _bootstrap_inner
  File "/usr/local/miniconda/lib/python3.7/threading.py", line 885 in _bootstrap

Current thread 0x00007f907ec38700 (most recent call first):
  File "/usr/local/miniconda/lib/python3.7/site-packages/nibabel/volumeutils.py", line 833 in _write_data
  File "/usr/local/miniconda/lib/python3.7/site-packages/nibabel/volumeutils.py", line 660 in array_to_file
  File "/usr/local/miniconda/lib/python3.7/site-packages/nibabel/arraywriters.py", line 562 in to_fileobj
  File "/usr/local/miniconda/lib/python3.7/site-packages/nibabel/analyze.py", line 1090 in to_file_map
  File "/usr/local/miniconda/lib/python3.7/site-packages/nibabel/filebasedimages.py", line 334 in to_filename
  File "/usr/local/miniconda/lib/python3.7/site-packages/niworkflows/interfaces/bids.py", line 431 in _run_interface
  File "/usr/local/miniconda/lib/python3.7/site-packages/nipype/interfaces/base/core.py", line 375 in run
  File "/usr/local/miniconda/lib/python3.7/site-packages/nipype/pipeline/engine/nodes.py", line 637 in _run_command
  File "/usr/local/miniconda/lib/python3.7/site-packages/nipype/pipeline/engine/nodes.py", line 557 in _run_interface
  File "/usr/local/miniconda/lib/python3.7/site-packages/nipype/pipeline/engine/nodes.py", line 473 in run
  File "/usr/local/miniconda/lib/python3.7/site-packages/nipype/pipeline/plugins/multiproc.py", line 316 in _send_procs_to_workers
  File "/usr/local/miniconda/lib/python3.7/site-packages/nipype/pipeline/plugins/base.py", line 183 in run
  File "/usr/local/miniconda/lib/python3.7/site-packages/nipype/pipeline/engine/workflows.py", line 599 in run
  File "/usr/local/miniconda/lib/python3.7/site-packages/fmriprep/cli/run.py", line 436 in main
  File "/usr/local/miniconda/bin/fmriprep", line 11 in <module>
fMRIPrep: Please report errors to https://github.com/poldracklab/fmriprep/issues


Thank you in advance for your answer!

**Comment by effigies** [comment]
Hi @marloula, this error being addressed in poldracklab/niworkflows#378. Apologies for the delay.
